src/boot/linux.c

```diff
@@ -119,16 +46,63 @@ EFI_STATUS linux_exec(
         if (err != EFI_SUCCESS)
                 return log_error_status(err, "Bad kernel image: %m");
 
-        _cleanup_(unload_imagep) EFI_HANDLE kernel_image = NULL;
-        err = load_image(parent, kernel->iov_base, kernel->iov_len, &kernel_image);
-        if (err != EFI_SUCCESS)
-                return log_error_status(err, "Error loading kernel image: %m");
+        const PeSectionHeader *headers;
+        size_t n_headers;
+
+        /* Do we need to validate anyting here? the len? */
+        err = pe_section_table_from_base(kernel->iov_base, &headers, &n_headers);
+        if (err != EFI_SUCCESS) {
+                return log_error_status(err, "Cannot read sections: %m");
+        }
+
+        /* Do we need to ensure under 4gb address on x86? */
+        _cleanup_pages_ Pages loaded_kernel_pages = xmalloc_pages(
+                        AllocateAnyPages, EfiLoaderCode, EFI_SIZE_TO_PAGES(kernel_size_in_memory), 0);
+
+        uint8_t* loaded_kernel = (uint8_t*)PHYSICAL_ADDRESS_TO_POINTER(loaded_kernel_pages.addr);
+        for (size_t i = 0; i < n_headers; i++) {
+                if (headers[i].PointerToRelocations != 0)
+                        return log_error_status(EFI_LOAD_ERROR, "Relocations unexpected");
+                if (headers[i].SizeOfRawData) {
```

Comment:
we generally prefer explicit non-zero comparisons for numeric values (unless they are conceptually booleans or boolean-like). or in other words, please write as `if (….SizeOfRawData != 0)`